File: test/matlab/rc4.m
%--------------------------------------------------------------------------
%   NAME:   rc4.m
%   Author: Hoang Phuc Nguyen - intern
%   Date:   19th Mar, 2018
%   Description: Use this script to test simulation result made by modelsim
%--------------------------------------------------------------------------


a = '../data/output.txt';
fileOut  = fopen(a, 'w');
fileIn   = fopen('../data/input.txt', 'w');

% A = fscanf(fileIn, '%x')'
keyLength = length;
% keyLength = A(1)
% key = A(2:A(1) + 1)

fprintf(fileIn, "%x\n", length);
key = randi(255, length, 1);

for i=1:length
   fprintf(fileIn, "%x\n", key(i)); 
end

for i = 0:255
   sBox(i + 1) = i;
end

j = 0;

for i = 0:255
   j = mod(j + sBox(i + 1) + key(mod(i, keyLength) + 1), 256)
   temp = sBox(i + 1);
   sBox(i + 1) = sBox(j + 1);
   sBox(j + 1) = temp;
end

sBox

n = 0;
j = 0;

for i = 1: keyLength
    j = mod(j + sBox(i + 1), 256);
    temp = sBox(i + 1);
    sBox(i + 1) = sBox(j + 1);
    sBox(j + 1) = temp;
    k = mod(sBox(i + 1) + sBox(j + 1), 256);
    Ckey(n + 1) = sBox(k + 1);
    fprintf(fileOut, "%x\n", Ckey(n + 1));
    n = n + 1;
end

File: test/c/rc4.cpp
#include <stdio.h>
#include <iostream>
#include <fstream>

using namespace std;

int main()
{
    ifstream fileIn;
    ofstream fileOut;
    int keyLength = 0;
    
    // read file from input.txt
    fileIn.open("input.txt");

    fileIn >> keyLength;
    cout << "key length: " << keyLength << endl;

    int *key = new int[keyLength];
    for (int i = 0; i < keyLength; i++)
    {
        int temp;
        fileIn >> temp;
        cout << temp << endl;
        key[i] = temp;
    }
    fileIn.close();

    fileOut.open("output.txt", ios::out);
    unsigned char sBox[256] = {0};

    unsigned int i = 0; 
    for (i = 0; i <= 255; i++)
    {
        sBox[i] = i;
    }

    // KSA
    unsigned int j = 0;
    for (i = 0; i <= 255; i++) 
    {
        j = (j + sBox[i] + key[i % keyLength]) % 256;
        int temp = sBox[i];
        sBox[i] = sBox[j];
        sBox[j] = temp;
        printf("%d\t", j);
        fileOut << (int)j << "\t";
        if (i % 10 == 0) 
        {
            fileOut << endl;
        }
    }
    fileOut << endl;

    printf("after KSA: \n");
    fileOut << "after KSA:" << endl;
    for (int i = 0; i < 256; i++) {
        printf("%d\t", sBox[i]);
        fileOut << (int)sBox[i] << "\t";
        if (i % 10 == 0) 
        {
           fileOut << endl;
        }
    }

    fileOut << endl;
    // PRGA
    j = 0;
    printf("key value: \n");
    fileOut << "Key value: " << endl;
    for (i = 1; i < keyLength + 1; i++) 
    {
        j = (j + sBox[i]) % 256;
        int temp = sBox[i];
        sBox[i] = sBox[j];
        sBox[j] = temp;
        int sk = (sBox[i] + sBox[j]) % 256;
        printf("%d\n", sBox[sk]);
        fileOut << (int)sBox[sk] << endl;
        if (i % 10 == 0) 
        {
            fileOut << endl;
        }
    }
    fileOut.close();
    return 0;
}